File: distribution/python/dqops/client/models/table_profiling_check_categories_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.profiling_time_period_truncation import ProfilingTimePeriodTruncation
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.table_accuracy_profiling_checks_spec import (
        TableAccuracyProfilingChecksSpec,
    )
    from ..models.table_availability_profiling_checks_spec import (
        TableAvailabilityProfilingChecksSpec,
    )
    from ..models.table_custom_sql_profiling_checks_spec import (
        TableCustomSqlProfilingChecksSpec,
    )
    from ..models.table_profiling_check_categories_spec_comparisons import (
        TableProfilingCheckCategoriesSpecComparisons,
    )
    from ..models.table_profiling_check_categories_spec_custom import (
        TableProfilingCheckCategoriesSpecCustom,
    )
    from ..models.table_schema_profiling_checks_spec import (
        TableSchemaProfilingChecksSpec,
    )
    from ..models.table_timeliness_profiling_checks_spec import (
        TableTimelinessProfilingChecksSpec,
    )
    from ..models.table_volume_profiling_checks_spec import (
        TableVolumeProfilingChecksSpec,
    )


T = TypeVar("T", bound="TableProfilingCheckCategoriesSpec")


@_attrs_define
class TableProfilingCheckCategoriesSpec:
    """
    Attributes:
        custom (Union[Unset, TableProfilingCheckCategoriesSpecCustom]): Dictionary of custom checks. The keys are check
            names within this category.
        result_truncation (Union[Unset, ProfilingTimePeriodTruncation]):
        volume (Union[Unset, TableVolumeProfilingChecksSpec]):
        timeliness (Union[Unset, TableTimelinessProfilingChecksSpec]):
        accuracy (Union[Unset, TableAccuracyProfilingChecksSpec]):
        custom_sql (Union[Unset, TableCustomSqlProfilingChecksSpec]):
        availability (Union[Unset, TableAvailabilityProfilingChecksSpec]):
        schema (Union[Unset, TableSchemaProfilingChecksSpec]):
        comparisons (Union[Unset, TableProfilingCheckCategoriesSpecComparisons]): Dictionary of configuration of checks
            for table comparisons. The key that identifies each comparison must match the name of a data comparison that is
            configured on the parent table.
    """

    custom: Union[Unset, "TableProfilingCheckCategoriesSpecCustom"] = UNSET
    result_truncation: Union[Unset, ProfilingTimePeriodTruncation] = UNSET
    volume: Union[Unset, "TableVolumeProfilingChecksSpec"] = UNSET
    timeliness: Union[Unset, "TableTimelinessProfilingChecksSpec"] = UNSET
    accuracy: Union[Unset, "TableAccuracyProfilingChecksSpec"] = UNSET
    custom_sql: Union[Unset, "TableCustomSqlProfilingChecksSpec"] = UNSET
    availability: Union[Unset, "TableAvailabilityProfilingChecksSpec"] = UNSET
    schema: Union[Unset, "TableSchemaProfilingChecksSpec"] = UNSET
    comparisons: Union[Unset, "TableProfilingCheckCategoriesSpecComparisons"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        custom: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.custom, Unset):
            custom = self.custom.to_dict()

        result_truncation: Union[Unset, str] = UNSET
        if not isinstance(self.result_truncation, Unset):
            result_truncation = self.result_truncation.value

        volume: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.volume, Unset):
            volume = self.volume.to_dict()

        timeliness: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.timeliness, Unset):
            timeliness = self.timeliness.to_dict()

        accuracy: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.accuracy, Unset):
            accuracy = self.accuracy.to_dict()

        custom_sql: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.custom_sql, Unset):
            custom_sql = self.custom_sql.to_dict()

        availability: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.availability, Unset):
            availability = self.availability.to_dict()

        schema: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.schema, Unset):
            schema = self.schema.to_dict()

        comparisons: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.comparisons, Unset):
            comparisons = self.comparisons.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if custom is not UNSET:
            field_dict["custom"] = custom
        if result_truncation is not UNSET:
            field_dict["result_truncation"] = result_truncation
        if volume is not UNSET:
            field_dict["volume"] = volume
        if timeliness is not UNSET:
            field_dict["timeliness"] = timeliness
        if accuracy is not UNSET:
            field_dict["accuracy"] = accuracy
        if custom_sql is not UNSET:
            field_dict["custom_sql"] = custom_sql
        if availability is not UNSET:
            field_dict["availability"] = availability
        if schema is not UNSET:
            field_dict["schema"] = schema
        if comparisons is not UNSET:
            field_dict["comparisons"] = comparisons

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.table_accuracy_profiling_checks_spec import (
            TableAccuracyProfilingChecksSpec,
        )
        from ..models.table_availability_profiling_checks_spec import (
            TableAvailabilityProfilingChecksSpec,
        )
        from ..models.table_custom_sql_profiling_checks_spec import (
            TableCustomSqlProfilingChecksSpec,
        )
        from ..models.table_profiling_check_categories_spec_comparisons import (
            TableProfilingCheckCategoriesSpecComparisons,
        )
        from ..models.table_profiling_check_categories_spec_custom import (
            TableProfilingCheckCategoriesSpecCustom,
        )
        from ..models.table_schema_profiling_checks_spec import (
            TableSchemaProfilingChecksSpec,
        )
        from ..models.table_timeliness_profiling_checks_spec import (
            TableTimelinessProfilingChecksSpec,
        )
        from ..models.table_volume_profiling_checks_spec import (
            TableVolumeProfilingChecksSpec,
        )

        d = src_dict.copy()
        _custom = d.pop("custom", UNSET)
        custom: Union[Unset, TableProfilingCheckCategoriesSpecCustom]
        if isinstance(_custom, Unset):
            custom = UNSET
        else:
            custom = TableProfilingCheckCategoriesSpecCustom.from_dict(_custom)

        _result_truncation = d.pop("result_truncation", UNSET)
        result_truncation: Union[Unset, ProfilingTimePeriodTruncation]
        if isinstance(_result_truncation, Unset):
            result_truncation = UNSET
        else:
            result_truncation = ProfilingTimePeriodTruncation(_result_truncation)

        _volume = d.pop("volume", UNSET)
        volume: Union[Unset, TableVolumeProfilingChecksSpec]
        if isinstance(_volume, Unset):
            volume = UNSET
        else:
            volume = TableVolumeProfilingChecksSpec.from_dict(_volume)

        _timeliness = d.pop("timeliness", UNSET)
        timeliness: Union[Unset, TableTimelinessProfilingChecksSpec]
        if isinstance(_timeliness, Unset):
            timeliness = UNSET
        else:
            timeliness = TableTimelinessProfilingChecksSpec.from_dict(_timeliness)

        _accuracy = d.pop("accuracy", UNSET)
        accuracy: Union[Unset, TableAccuracyProfilingChecksSpec]
        if isinstance(_accuracy, Unset):
            accuracy = UNSET
        else:
            accuracy = TableAccuracyProfilingChecksSpec.from_dict(_accuracy)

        _custom_sql = d.pop("custom_sql", UNSET)
        custom_sql: Union[Unset, TableCustomSqlProfilingChecksSpec]
        if isinstance(_custom_sql, Unset):
            custom_sql = UNSET
        else:
            custom_sql = TableCustomSqlProfilingChecksSpec.from_dict(_custom_sql)

        _availability = d.pop("availability", UNSET)
        availability: Union[Unset, TableAvailabilityProfilingChecksSpec]
        if isinstance(_availability, Unset):
            availability = UNSET
        else:
            availability = TableAvailabilityProfilingChecksSpec.from_dict(_availability)

        _schema = d.pop("schema", UNSET)
        schema: Union[Unset, TableSchemaProfilingChecksSpec]
        if isinstance(_schema, Unset):
            schema = UNSET
        else:
            schema = TableSchemaProfilingChecksSpec.from_dict(_schema)

        _comparisons = d.pop("comparisons", UNSET)
        comparisons: Union[Unset, TableProfilingCheckCategoriesSpecComparisons]
        if isinstance(_comparisons, Unset):
            comparisons = UNSET
        else:
            comparisons = TableProfilingCheckCategoriesSpecComparisons.from_dict(
                _comparisons
            )

        table_profiling_check_categories_spec = cls(
            custom=custom,
            result_truncation=result_truncation,
            volume=volume,
            timeliness=timeliness,
            accuracy=accuracy,
            custom_sql=custom_sql,
            availability=availability,
            schema=schema,
            comparisons=comparisons,
        )

        table_profiling_check_categories_spec.additional_properties = d
        return table_profiling_check_categories_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/checks/table/profiling/TableProfilingCheckCategoriesSpec.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.checks.table.profiling;

import com.dqops.checks.*;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.scheduling.CheckRunScheduleGroup;
import com.dqops.metadata.sources.TableSpec;
import com.dqops.metadata.timeseries.TimeSeriesConfigurationSpec;
import com.dqops.metadata.timeseries.TimeSeriesMode;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.dqops.utils.serialization.IgnoreEmptyYamlSerializer;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import lombok.EqualsAndHashCode;

import java.util.Objects;

/**
 * Container of table level checks that are activated on a table level.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class TableProfilingCheckCategoriesSpec extends AbstractRootChecksContainerSpec {
    public static final ChildHierarchyNodeFieldMapImpl<TableProfilingCheckCategoriesSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractRootChecksContainerSpec.FIELDS) {
        {
            put("volume", o -> o.volume);
            put("timeliness", o -> o.timeliness);
            put("accuracy", o -> o.accuracy);
            put("custom_sql", o -> o.customSql);
            put("availability", o -> o.availability);
            put("schema", o -> o.schema);
            put("comparisons", o -> o.comparisons);
        }
    };

    @JsonPropertyDescription("Defines how many profiling checks results are stored for the table monthly. By default, DQOps will use the 'one_per_month' configuration and store only the most recent " +
         "profiling checks result executed during the month. By changing this value, it is possible to store one value per day or even store all profiling checks results.")
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private ProfilingTimePeriod resultTruncation;

    @JsonPropertyDescription("Configuration of volume data quality checks on a table level.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableVolumeProfilingChecksSpec volume;

    @JsonPropertyDescription("Configuration of timeliness checks on a table level. Timeliness checks detect anomalies like rapid row count changes.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableTimelinessProfilingChecksSpec timeliness;

    @JsonPropertyDescription("Configuration of accuracy checks on a table level. Accuracy checks compare the tested table with another reference table.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableAccuracyProfilingChecksSpec accuracy;

    @JsonPropertyDescription("Configuration of data quality checks that are evaluating custom SQL conditions and aggregated expressions.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableCustomSqlProfilingChecksSpec customSql;

    @JsonPropertyDescription("Configuration of the table availability data quality checks on a table level.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableAvailabilityProfilingChecksSpec availability;

    @JsonPropertyDescription("Configuration of schema (column count and schema) data quality checks on a table level.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableSchemaProfilingChecksSpec schema;

    @JsonPropertyDescription("Dictionary of configuration of checks for table comparisons. The key that identifies each comparison must match the name of a data comparison that is configured on the parent table.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableComparisonProfilingChecksSpecMap comparisons = new TableComparisonProfilingChecksSpecMap();

    /**
     * Returns the result truncation configuration.
     * @return Result truncation policy.
     */
    public ProfilingTimePeriod getResultTruncation() {
        return resultTruncation;
    }

    /**
     * Sets the result truncation configuration.
     * @param resultTruncation New result truncation configuration.
     */
    public void setResultTruncation(ProfilingTimePeriod resultTruncation) {
        this.setDirtyIf(!Objects.equals(this.resultTruncation, resultTruncation));
        this.resultTruncation = resultTruncation;
    }

    /**
     * Returns a container of volume check configuration on a table level.
     * @return Volume checks configuration.
     */
    public TableVolumeProfilingChecksSpec getVolume() {
        return volume;
    }

    /**
     * Sets a reference to a volume checks container.
     * @param volume New volume checks configuration.
     */
    public void setVolume(TableVolumeProfilingChecksSpec volume) {
        this.setDirtyIf(!Objects.equals(this.volume, volume));
        this.volume = volume;
        this.propagateHierarchyIdToField(volume, "volume");
    }

    /**
     * Returns a container of timeliness checks.
     * @return Timeliness data quality checks on a table level configuration.
     */
    public TableTimelinessProfilingChecksSpec getTimeliness() {
        return timeliness;
    }

    /**
     * Sets a reference to a timeliness table level checks container.
     * @param timeliness New timeliness checks.
     */
    public void setTimeliness(TableTimelinessProfilingChecksSpec timeliness) {
        this.setDirtyIf(!Objects.equals(this.timeliness, timeliness));
        this.timeliness = timeliness;
        this.propagateHierarchyIdToField(timeliness, "timeliness");
    }

    /**
     * Returns the configuration of accuracy checks.
     * @return Accuracy checks.
     */
    public TableAccuracyProfilingChecksSpec getAccuracy() {
        return accuracy;
    }

    /**
     * Sets the configuration of accuracy checks.
     * @param accuracy Accuracy checks.
     */
    public void setAccuracy(TableAccuracyProfilingChecksSpec accuracy) {
        this.setDirtyIf(!Objects.equals(this.accuracy, accuracy));
        this.accuracy = accuracy;
        this.propagateHierarchyIdToField(accuracy, "accuracy");
    }

    /**
     * Returns a container of custom sql checks.
     * @return Custom sql checks.
     */
    public TableCustomSqlProfilingChecksSpec getCustomSql() {
        return customSql;
    }

    /**
     * Sets a reference to a custom sql checks container.
     * @param customSql Custom sql checks.
     */
    public void setCustomSql(TableCustomSqlProfilingChecksSpec customSql) {
        this.setDirtyIf(!Objects.equals(this.customSql, customSql));
        this.customSql = customSql;
        this.propagateHierarchyIdToField(customSql, "custom_sql");
    }

    /**
     * Returns a container of the table availability checks.
     * @return Table availability checks.
     */
    public TableAvailabilityProfilingChecksSpec getAvailability() {
        return availability;
    }

    /**
     * Sets a reference to an availability checks container.
     * @param availability Table availability checks.
     */
    public void setAvailability(TableAvailabilityProfilingChecksSpec availability) {
        this.setDirtyIf(!Objects.equals(this.availability, availability));
        this.availability = availability;
        this.propagateHierarchyIdToField(availability, "availability");
    }

    /**
     * Returns a container of table schema checks.
     * @return Table schema checks.
     */
    public TableSchemaProfilingChecksSpec getSchema() {
        return schema;
    }

    /**
     * Sets the reference to the table schema checks.
     * @param schema Table schema checks container.
     */
    public void setSchema(TableSchemaProfilingChecksSpec schema) {
        this.setDirtyIf(!Objects.equals(this.schema, schema));
        this.schema = schema;
        this.propagateHierarchyIdToField(schema, "schema");
    }

    /**
     * Returns the dictionary of comparisons.
     * @return Dictionary of comparisons.
     */
    @Override
    public TableComparisonProfilingChecksSpecMap getComparisons() {
        return comparisons;
    }

    /**
     * Sets the dictionary of comparisons.
     * @param comparisons Dictionary of comparisons.
     */
    public void setComparisons(TableComparisonProfilingChecksSpecMap comparisons) {
        this.setDirtyIf(!Objects.equals(this.comparisons, comparisons));
        this.comparisons = comparisons;
        this.propagateHierarchyIdToField(comparisons, "comparisons");
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Returns time series configuration for the given group of checks.
     *
     * @param tableSpec Parent table specification - used to get the details about the time partitioning column.
     * @return Time series configuration.
     */
    @Override
    public TimeSeriesConfigurationSpec getTimeSeriesConfiguration(TableSpec tableSpec) {
        ProfilingTimePeriod profilingTimePeriod = this.resultTruncation != null ? this.resultTruncation : ProfilingTimePeriod.one_per_month;

        return new TimeSeriesConfigurationSpec()
        {{
            setMode(TimeSeriesMode.current_time);
            setTimeGradient(profilingTimePeriod.toTimePeriodGradient());
        }};
    }

    /**
     * Returns the type of checks (profiling, monitoring, partitioned).
     *
     * @return Check type.
     */
    @Override
    @JsonIgnore
    public CheckType getCheckType() {
        return CheckType.profiling;
    }

    /**
     * Returns the time range for monitoring and partitioned checks (daily, monthly, etc.).
     * Profiling checks do not have a time range and return null.
     *
     * @return Time range (daily, monthly, ...).
     */
    @Override
    @JsonIgnore
    public CheckTimeScale getCheckTimeScale() {
        return null;
    }

    /**
     * Returns the check target, where the check could be applied.
     *
     * @return Check target, "table" or "column".
     */
    @Override
    @JsonIgnore
    public CheckTarget getCheckTarget() {
        return CheckTarget.table;
    }

    /**
     * Returns the name of the cron expression that is used to schedule checks in this check root object.
     *
     * @return Monitoring schedule group (named schedule) that is used to schedule the checks in this root.
     */
    @Override
    @JsonIgnore
    public CheckRunScheduleGroup getSchedulingGroup() {
        return CheckRunScheduleGroup.profiling;
    }

    public static class TableProfilingCheckCategoriesSpecSampleFactory implements SampleValueFactory<TableProfilingCheckCategoriesSpec> {
        @Override
        public TableProfilingCheckCategoriesSpec createSample() {
            return new TableProfilingCheckCategoriesSpec() {{
                setVolume(new TableVolumeProfilingChecksSpec.TableVolumeProfilingChecksSpecSampleFactory().createSample());
                setComparisons(null);
            }};
        }
    }
}
